main/src/ca/uwaterloo/flix/language/errors/ParseError.scala

```diff
@@ -18,28 +18,234 @@ package ca.uwaterloo.flix.language.errors
 
 import ca.uwaterloo.flix.language.CompilationMessage
 import ca.uwaterloo.flix.language.ast.Ast.SyntacticContext
-import ca.uwaterloo.flix.language.ast.SourceLocation
-import ca.uwaterloo.flix.util.Formatter
+import ca.uwaterloo.flix.language.ast.{SourceLocation, TokenKind}
+import ca.uwaterloo.flix.util.{Formatter, InternalCompilerException}
 
 /**
- * An error raised to indicate a parse error.
- *
- * @param msg the error message.
- * @param ctx the syntactic context.
- * @param loc the source location.
- */
-case class ParseError(msg: String, ctx: SyntacticContext, loc: SourceLocation) extends CompilationMessage with Recoverable {
+  * A common super-type for parser errors.
+  */
+sealed trait ParseError extends CompilationMessage {
   val kind = "Parse Error"
+  val context: SyntacticContext = SyntacticContext.Unknown
+}
+
+object ParseError {
+
+  /**
+    * Joins items nicely with comma separation ending with an "or".
+    * For instance prettyJoin(List("def", "enum", "trait")) gives "def, enum or trait".
+    */
+  private def prettyJoin[T](items: Seq[T]): String = items match {
+    case i1 :: i2 :: Nil => s"$i1 or $i2"
+    case i1 :: Nil => s"$i1"
+    case i :: tail => s"$i, ${prettyJoin(tail)}"
+  }
+
+  /**
+    * An error raised to indicate an unexpected token was found.
+    *
+    * @param expected Names of the tokens that are expected at the location. See [[TokenKind.display]].
+    * @param actual   Name of the token that was actually found. See [[TokenKind.display]].
+    * @param ctx      The syntactic context.
+    * @param loc      The source location.
+    * @param hint     Optional hint with more details about the error
+    */
+  case class UnexpectedToken(expected: Seq[String], actual: Option[String], ctx: SyntacticContext, loc: SourceLocation, hint: Option[String] = None) extends ParseError with Recoverable {
+    override val kind = s"Parse Error ($ctx)"
+    override val context: SyntacticContext = ctx
+
+    def summary: String = {
+      val expectedStr = s"Expected ${prettyJoin(expected)}"
+      val actualStr = actual.map(a => s" before $a").getOrElse("")
+      s"$expectedStr$actualStr."
+    }
+
+    def message(formatter: Formatter): String = {
```

Comment:
I see your point. The "errors as data-types" makes a lot of sense when we consider IDE-integration also 👍

To me it seems like in this particular case we are pretty much doing that? The string is being built in two-stages, but it's still being built from data that should be generally applicable.

Originally I passed in `TokenKind`s rather than strings, but it doesn't cover all the use-cases. For instance, we want to report that we `Expected <expression>` instead of a list of all the tokens that can lead an expression.

Reply:
Yes. This is the good kind of stuff I am looking for! 🚀